ffmpeg.c

```diff
@@ -767,6 +786,218 @@ static void write_frame(AVFormatContext *s, AVPacket *pkt, OutputStream *ost)
     av_packet_unref(pkt);
 }
 
+static void *ff_encode_video_thread(void *args)
+{
+    EncodeThreadContext *ctx = (EncodeThreadContext *)args;
+    if (!ctx)
+        return NULL;
+    while (1) {
+        if (!ctx->process_data_cb)
+            break;
+        pthread_mutex_lock(&ctx->in_queue_lock);
+        if (ctx->in_queue.available == 0) {
+            if (ctx->status == ENCODE_THREAD_GOT_EOS) {
+                pthread_mutex_unlock(&ctx->in_queue_lock);
+                ctx->status = ENCODE_THREAD_FLUSH_OUT;
+            }
+            pthread_cond_wait(&ctx->in_cond, &ctx->in_queue_lock); // wait the packet to decode
+            if (ctx->status == ENCODE_THREAD_EXIT)
+                break;
+            pthread_mutex_unlock(&ctx->in_queue_lock);
+            usleep(100);
+            continue;
+        }
+        AVFrame *frame = ff_bufqueue_get(&ctx->in_queue);
+        pthread_mutex_unlock(&ctx->in_queue_lock);
+        ctx->process_data_cb (ctx->params, frame);
+        if (ctx->status == ENCODE_THREAD_EXIT)
+            break;
+    }
+    pthread_mutex_lock(&ctx->priv_lock);
+    ctx->status = ENCODE_THREAD_NOT_INIT;
+    pthread_mutex_unlock(&ctx->priv_lock);
+    return NULL;
+}
+
+static int ff_encode_video_push_data(EncodeThreadContext *ctx, AVFrame *frame)
+{
+    while (ctx->status < ENCODE_THREAD_GOT_EOS) {
+        /* need enque eos buffer more than once */
+        pthread_mutex_lock(&ctx->in_queue_lock);
+        if (ctx->in_queue.available < ENCODE_QUEUE_SIZE) {
+            ff_bufqueue_add(NULL, &ctx->in_queue, frame);
+            pthread_cond_signal(&ctx->in_cond);
+            pthread_mutex_unlock(&ctx->in_queue_lock);
+            break;
+        }
+        pthread_mutex_unlock(&ctx->in_queue_lock);
+        usleep(100);
+    };
+    return 0;
+}
+
+static int ff_encode_video_thread_create(EncodeThreadContext *ctx, AVFormatContext *s, OutputStream *ost)
+{
+    pthread_mutex_lock(&ctx->priv_lock);
+    switch (ctx->status) {
+        case ENCODE_THREAD_EXIT:
+        case ENCODE_THREAD_NOT_INIT:
+            ctx->status = ENCODE_THREAD_RUNING;
+            ctx->params->s = s;
+            ctx->params->ost = ost;
+            pthread_create(&ctx->thread_id, NULL, ff_encode_video_thread, ctx);
+            break;
+        case ENCODE_THREAD_RUNING:
+            break;
+        case ENCODE_THREAD_GOT_EOS:
+            pthread_cond_signal(&ctx->in_cond);
+            break;
+        default:
+            break;
+    }
+    pthread_mutex_unlock(&ctx->priv_lock);
+    return 0;
+}
+
+static EncodeThreadStatus ff_encode_video_read_status(EncodeThreadContext *ctx)
+{
+    if (!ctx)
+        return ENCODE_THREAD_NOT_INIT;
+    EncodeThreadStatus status;
+    pthread_mutex_lock(&ctx->priv_lock);
+    status = ctx->status;
+    pthread_mutex_unlock(&ctx->priv_lock);
+    return status;
+}
+
+static int ff_encode_video_set_stream_eof(EncodeThreadContext *ctx)
```

Comment:
Merge the status update function like:
ff_encode_video_set_stream_status(EncodeThreadContext *ctx, EncodeThreadStatus status;)